api/v1/recorder/recorder.proto

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
+
+package recorder;
+
+option go_package = "github.com/cilium/cilium/api/v1/recorder";
+
+// Recorder implements the Hubble module for capturing network packets
+service Recorder {
+    // Record can start and stop a single recording. The recording is
+    // automatically stopped if the client aborts this rpc call.
+    rpc Record (stream RecordRequest) returns (stream RecordResponse);
+}
+
+message RecordRequest {
+    oneof request_type {
+        // start starts a new recording with the given parameters.
+        StartRecording start = 1;
+        // stop stops the running recording.
+        StopRecording stop = 2;
+    }
+}
+
+message StartRecording {
+    // filesink configures the outfile of this recording
+    // Future alternative sink configurations may be added as a
+    // backwards-compatible change by moving this field into a oneof.
+    FileSinkConfiguration filesink = 1;
+
+    // include list for this recording. Packets matching any of the provided
+    // filters will be recorded.
+    repeated Filter include = 2;
+
+    // max_capture_length specifies the maximum packet length.
+    // Full packet length will be captured if absent/zero.
+    uint32 max_capture_length = 3;
+}
+
+// FileSinkConfiguration configures the file output. Possible future additions
+// might be the selection of the output volume. The initial implementation will
+// only support a single volume which is configured as a cilium-agent CLI flag.
+message FileSinkConfiguration {
+    // file_prefix is an optional prefix for the file name.
+    // Defaults to `hubble` if empty. Must match the following regex if not
+    // empty: ^[a-z][a-z0-9]{0,19}$
+    // The generated filename will be of format
+    //  <file_prefix>_<unixtime>_<unique_random>_<node_name>.pcap
+    string file_prefix = 1;
+}
+
+message Filter {
+    // source_cidr. Must not be empty.
+    // Set to 0.0.0.0/0 to match any IPv4 source address (::/0 for IPv6).
+    string source_cidr = 1;
```

Comment:
I'm asking because this looks like it becomes API. How would we extend this in the future for ARP or non-{TCP/UDP} traffic. I assume zero these fields?

Reply:
So the current API is intentionally very restrictive in that it requires the user to explicitly pick between IPv4 or IPv6. This is why I don't default to zero, because it would not be clear if the user means IPv4 or IPv6 "zero". If the user wants any IPv4 and any IPv6 traffic captured, they need to submit two filters to make that explicit. We can always relax that in the future but I'd prefer for the initial version to be a 1:1 mapping to the datapath filters.

You do bring up a good point in that I have not considered how this API would look like for non-IP traffic, because it does mean that an empty value here could either mean "match any ip traffic" or "match only non-ip traffic". Maybe we should wrap the L3 and L4 fields separately, such that a layer can explicitly be absent will well-defined semantics. I'll play around to see how complicated that would be.

Reply:
@gandro @jrfastab Hmm, right now at the point where we add the `cilium_capture_{in,out}()` we will never see non-ip traffic. But I guess there could be a l3 filter of some sort, so maybe the api could be extended with an proto_l3 field, not sure yet hmm.

Reply:
Yes, thinking about it some more, I think the current API should be future proof enough. We can relax it similarly to how the normal Hubble `FlowFilter` works, i.e. where a absent field just means "don't care about this, check if any other field is applicable". This should give us enough flexibility to support adding new fields to e.g. filter ARP traffic.

Reply:
Agree, sounds reasonable to me.